File: scripts/generate-icons.mjs
// One-off PWA icon generator. Run once with: `node scripts/generate-icons.mjs`
//
// The brand logo (public/images/saakie.jpg) is a 1080x1080 square image with no
// alpha channel. Each icon is produced by centering the logo (resized with
// `fit: contain`) onto a square brand near-black (#161616) canvas. The maskable
// icon keeps the logo inside the safe circle (~70% of the tile) so platforms that
// crop to a circle never clip it.
//
// `sharp` ships transitively with Next.js, so no extra dependency is required.
import sharp from 'sharp';
import { mkdir } from 'node:fs/promises';

const SRC = 'public/images/saakie.jpg';
const ICONS_DIR = 'public/icons';
const BG = { r: 0x16, g: 0x16, b: 0x16, alpha: 1 }; // brand near-black #161616

await mkdir(ICONS_DIR, { recursive: true });

// Place the logo, scaled to `logoFraction` of the tile, centered on a square BG.
async function squareIcon(size, logoFraction, outPath) {
  const inner = Math.round(size * logoFraction);
  const logo = await sharp(SRC)
    .resize(inner, inner, { fit: 'contain', background: { r: 0, g: 0, b: 0, alpha: 0 } })
    .png()
    .toBuffer();

  await sharp({
    create: { width: size, height: size, channels: 4, background: BG },
  })
    .composite([{ input: logo, gravity: 'center' }])
    .png({ compressionLevel: 9 })
    .toFile(outPath);
}

// Manifest icons (referenced from app/manifest.ts via /icons/*).
await squareIcon(192, 0.92, `${ICONS_DIR}/icon-192.png`);
await squareIcon(512, 0.92, `${ICONS_DIR}/icon-512.png`);
// Maskable: content within the ~70% safe circle.
await squareIcon(512, 0.7, `${ICONS_DIR}/maskable-512.png`);
// Apple touch icon (Apple applies its own rounding).
await squareIcon(180, 0.86, `${ICONS_DIR}/apple-icon-180.png`);

// Next.js file-convention head icons. These let Next inject the correct
// <link rel="icon"> / apple-touch-icon tags automatically from app/.
await squareIcon(512, 0.92, 'app/icon.png');
await squareIcon(180, 0.86, 'app/apple-icon.png');

console.log('Generated PWA icons in public/icons/, app/icon.png, app/apple-icon.png');

File: scripts/generate-icons.py
#!/usr/bin/env python3
"""Generate the PWA / home-screen icons from the side-menu wordmark.

Run from the repo root:  python3 scripts/generate-icons.py

Why this exists, and why it centres the way it does
---------------------------------------------------
`public/images/saakieLogo.png` is the wordmark the side menu shows. Despite the
extension it is a JPEG, 14173x7087, with a black background baked in and a
maroon swash sweeping behind the letters.

Centring on the image's pixel bounds puts the wordmark visibly high and left,
because the swash reaches further down and left than the letters do:

    content incl. swash (lum > 10)   centre (6612, 3910)
    the letters alone   (lum > 90)   centre (7040, 3466)   <- what the eye reads

So the letters are isolated by brightness and *their* centre is what gets
pinned to the middle of the tile. The rest of the source is pasted along with
them, so the swash still appears — it simply no longer drags the composition
off centre. Anything past the tile edge is cropped, which is invisible because
the source background and the canvas are both pure black.

Pure black is also the manifest's `background_color`, so the icon, its splash
screen and the letterboxing all agree and there is no seam.

Requires Pillow (`pip install Pillow`). The previous Node version of this
script was removed: it pointed at `public/images/saakie.jpg`, which no longer
exists in the repo, and `sharp` is not resolvable from this project.
"""

# Builtin generics in annotations (`tuple[int, ...]`) need this on Python 3.8,
# which is what this machine runs.
from __future__ import annotations

from pathlib import Path

from PIL import Image

Image.MAX_IMAGE_PIXELS = None

SRC = Path("public/images/saakieLogo.png")
ICONS_DIR = Path("public/icons")

# Brightness above which a pixel counts as part of the wordmark rather than the
# swash or JPEG noise. 90 and 140 give an identical box, so the letters are
# cleanly separated from everything else and the exact value is not delicate.
LETTER_THRESHOLD = 90

BACKGROUND = (0, 0, 0)


def letter_box(image: Image.Image) -> tuple[int, int, int, int]:
    """Bounding box of the white lettering, ignoring the swash behind it."""
    mask = image.convert("L").point(lambda v: 255 if v > LETTER_THRESHOLD else 0)
    box = mask.getbbox()
    if box is None:
        raise SystemExit(f"No lettering found in {SRC} above luminance {LETTER_THRESHOLD}")
    return box


def render(source: Image.Image, box, size: int, letter_width: float, out: Path) -> None:
    """Square icon with the lettering `letter_width` wide and dead centre.

    `letter_width` is a fraction of the tile: it constrains the *letters*, not
    the whole source image, which is what keeps every icon optically the same
    weight regardless of how much swash happens to surround it.
    """
    left, top, right, bottom = box
    scale = (size * letter_width) / (right - left)

    scaled = source.resize(
        (max(1, round(source.width * scale)), max(1, round(source.height * scale))),
        Image.LANCZOS,
    )

    # Where the letters' centre landed after scaling, and the offset that moves
    # it to the middle of the tile.
    centre_x = ((left + right) / 2) * scale
    centre_y = ((top + bottom) / 2) * scale

    canvas = Image.new("RGB", (size, size), BACKGROUND)
    canvas.paste(scaled, (round(size / 2 - centre_x), round(size / 2 - centre_y)))

    out.parent.mkdir(parents=True, exist_ok=True)
    canvas.save(out, "PNG", optimize=True)
    print(f"  {out}  ({size}x{size}, letters {letter_width:.0%} wide)")


def main() -> None:
    if not SRC.exists():
        raise SystemExit(f"Missing source image: {SRC}")

    source = Image.open(SRC).convert("RGB")
    box = letter_box(source)
    print(f"Source {source.size}, lettering {box[2] - box[0]}x{box[3] - box[1]} at {box}")

    # Regular icons: the wordmark is long, so 78% still reads comfortably while
    # leaving margin on the sides.
    render(source, box, 192, 0.78, ICONS_DIR / "icon-192.png")
    render(source, box, 512, 0.78, ICONS_DIR / "icon-512.png")

    # Maskable: Android may crop to a circle inscribed in the middle 80% of the
    # tile. At the wordmark's ~4:1 ratio, 60% wide keeps every corner of the
    # lettering inside that circle with room to spare.
    render(source, box, 512, 0.60, ICONS_DIR / "maskable-512.png")

    # Apple rounds the corners itself and never crops to a circle.
    render(source, box, 180, 0.76, ICONS_DIR / "apple-icon-180.png")

    # Next.js file conventions — these produce the <link rel="icon"> tags.
    render(source, box, 512, 0.78, Path("app/icon.png"))
    render(source, box, 180, 0.76, Path("app/apple-icon.png"))


if __name__ == "__main__":
    main()
